### PR TITLE
simple-cipher: Add canonical data

### LINKED ISSUE #586: simple-cipher: Implement canonical-data.json
We want to have a standard set of test inputs and outputs for each exercise to
make it easier to port them to new languages, as well as to help keep the
exercises in sync.

The Simple Cipher exercise can be found in the [./exercises/simple-cipher/][exercise-dir] directory.

The step-by-step instructions for how to compile a canonical-data.json file is
described in [this section of the contributing guide][contributing].

See http://exercism.io/contribute/canonical-data/simple-cipher for the up-to-date list of
language tracks that have an implementation of the Simple Cipher exercise.

The [tracking issue][tracking-issue] has more context.

[contributing]: https://github.com/exercism/docs/blob/master/you-can-help/improve-exercise-metadata.md#extracting-canonical-test-data
[tracking-issue]: https://github.com/exercism/x-common/issues/552
[exercise-dir]: https://github.com/exercism/x-common/tree/master/exercises/simple-cipher

### PR DESCRIPTION
**Pull request by aos**
Closes #586 

Hello. I've been working on this one. There are a few places where I wasn't sure what to do, specifically L27, and others where you have to do multiple encode into decodes. Or when adding a new cipher key. Let me know what you think!

### CODE REVIEW

**Comment by m-a-ge** [comment]
@aos build failed due to:
```text
exercises/simple-cipher/canonical-data.json invalid
[ { keyword: 'oneOf',
 dataPath: '.cases[0]',
 schemaPath: '#/oneOf',
 params: {},
 message: 'should match exactly one schema in oneOf' } ]
```

**Comment by Insti** [comment]
Link to problem description: https://github.com/exercism/problem-specifications/blob/master/exercises/simple-cipher/description.md

**Comment by Insti** [comment]
Hi @aos have you had a chance to look into why the json file is not valid?

**Comment by aos** [comment]
Hey @Insti. I looked at it briefly but couldn't understand. Sorry I have been very busy with work. I will look at it more tonight when I get a chance!

**Comment by Insti** [comment]
Try pasting the json into a schema checker for some clues:

https://jsonschemalint.com/#/version/draft-04/markup/json?gist=ba6a1e74397ade6bdbff65a9df0d9d3f

**Comment by aos** [comment]
Oh... that's handy. I'll look through it in a bit. Thank you!

**Comment by ErikSchierboom** [comment]
@aos Did you find any time to look at the problem? Do you perhaps need any help?

**Comment by aos** [comment]
@ErikSchierboom @Insti Hey finally got around to it. The latest commit fixes the schema validation issue. Let me know what still needs work -- I'll eventually squash these into a single commit

**Comment by coriolinus** [comment]
> you can assign whatever attributes you like to the expected object

cf. the Readme:

> There are also some convention about expected that you must follow:
> - if the input is valid but there is no result for the input, the value at "expected" should be null.
> - if an error is expected (because the input is invalid, or any other reason), the value at "expected" should be an object containing exactly one property, "error", whose value is a string.
>    - The string should explain why the error would occur.
>    - A particular track's implementation of the exercise need not necessarily check that the error includes that exact string as the cause, depending on what is idiomatic in the language (it may not be idiomatic to check strings for error messages).

File: exercises/simple-cipher/canonical-data.json [hunk 1]
```diff
@@ -0,0 +1,123 @@
+{
+  "exercise": "simple-cipher",
+  "version": "1.0.0",
+  "cases": [
+    {
+      "description": "Random key cipher",
+      "property": "random",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: I'm not sure, but I don't think this line actually does anything. The `property` key is normally used for specifying which method/property to test. As this is just a container for individual cases, I think it can be removed.

File: exercises/simple-cipher/canonical-data.json [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+          "description": "It is reversible",
+          "property": "encode -> decode",
+          "input_encode": "plaintext string",
+          "output_decode": "plaintext string"
+        }
+      ]
+    },
+    {
+      "description": "Incorrect key cipher",
+      "property": "incorrect",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: I'm not sure, but I don't think this line actually does anything. The `property` key is normally used for specifying which method/property to test. As this is just a container for individual cases, I think it can be removed.

File: exercises/simple-cipher/canonical-data.json [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          "description": "It throws an error with empty key",
+          "property": "Cipher Key",
+          "input_empty_string": "''",
+          "output_error": "Bad key"
+        }
+      ]
+    },
+    {
+      "description": "Substitution cipher",
+      "property": "substitution",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: I'm not sure, but I don't think this line actually does anything. The `property` key is normally used for specifying which method/property to test. As this is just a container for individual cases, I think it can be removed.

File: exercises/simple-cipher/canonical-data.json [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  "cases": [
+    {
+      "description": "Random key cipher",
+      "property": "random",
+      "cases": [
+        {
+          "description": "It has a key made of letters",
+          "property": "key",
+          "input": "any",
+          "matches_regex": "^[a-z]+$"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: We are in the process of enforcing `lowerCamelCase` keys (see #987), so I would suggest to rename the keys (not only this one) to `lowerCamelCase`.

File: exercises/simple-cipher/canonical-data.json [hunk 5]
```diff
@@ -0,0 +1,123 @@
+{
+  "exercise": "simple-cipher",
+  "version": "1.0.0",
+  "cases": [
+    {
+      "description": "Random key cipher",
+      "property": "random",
+      "cases": [
+        {
+          "description": "It has a key made of letters",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Maybe shorten to "Key has only lower-case letters"?

File: exercises/simple-cipher/canonical-data.json [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        {
+          "description": "It can decode",
+          "property": "decode",
+          "input_key": "cipher.key",
+          "output_string": "aaaaaaaaaa"
+        },
+        {
+          "description": "It is reversible",
+          "property": "encode -> decode",
+          "input_encode": "plaintext string",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Maybe rename to `plaintext` or `input`?

File: exercises/simple-cipher/canonical-data.json [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+          "description": "It can encode",
+          "property": "encode",
+          "input_plaintext": "aaaaaaaaaa",
+          "output_key": "cipher.key"
+        },
+        {
+          "description": "It can decode",
+          "property": "decode",
+          "input_key": "cipher.key",
+          "output_string": "aaaaaaaaaa"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: This should be renamed to `expected`.

File: exercises/simple-cipher/canonical-data.json [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It is reversible",
+          "property": "encode -> decode",
+          "input_encode": "plaintext string",
+          "output_decode": "plaintext string"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: This should be renamed to `expected`.

File: exercises/simple-cipher/canonical-data.json [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    },
+    {
+      "description": "Incorrect key cipher",
+      "property": "incorrect",
+      "cases": [
+        {
+          "description": "It throws an error with an all caps key",
+          "property": "Cipher Key",
+          "input_all_caps_string": "ABCDEF",
+          "output_error": "Bad key"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: This property should be renamed to `expected` and should be structured as an object with a single `error` key. See [this example](https://github.com/exercism/problem-specifications/blob/master/exercises/all-your-base/canonical-data.json#L150).

File: exercises/simple-cipher/canonical-data.json [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It throws an error with a numeric key",
+          "property": "Cipher Key",
+          "input_all_number_string": "'12345'",
+          "output_error": "Bad key"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: This property should be renamed to `expected` and should be structured as an object with a single `error` key. See [this example](https://github.com/exercism/problem-specifications/blob/master/exercises/all-your-base/canonical-data.json#L150).

File: exercises/simple-cipher/canonical-data.json [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It throws an error with empty key",
+          "property": "Cipher Key",
+          "input_empty_string": "''",
+          "output_error": "Bad key"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: This property should be renamed to `expected` and should be structured as an object with a single `error` key. See [this example](https://github.com/exercism/problem-specifications/blob/master/exercises/all-your-base/canonical-data.json#L150).

File: exercises/simple-cipher/canonical-data.json [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It throws an error with a numeric key",
+          "property": "Cipher Key",
+          "input_all_number_string": "'12345'",
+          "output_error": "Bad key"
+        },
+        {
+          "description": "It throws an error with empty key",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Maybe rephrase to something like: "An empty key is invalid"

File: exercises/simple-cipher/canonical-data.json [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+          "property": "Cipher Key",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Maybe rename the property to `key`?

File: exercises/simple-cipher/canonical-data.json [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+          "input_empty_string": "''",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: I would rename to `input`.

File: exercises/simple-cipher/canonical-data.json [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+            "property": "encode",
+            "input_string": "iamapandabear",
+            "output_string": "qayaeaagacia"
+          }
+        },
+        {
+          "description": "It can wrap on encode",
+          "property": "encode",
+          "input_string": "zzzzzzzzzz",
+          "output_string": "zabcdefghi"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `expected`

File: exercises/simple-cipher/canonical-data.json [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It can wrap on decode",
+          "property": "decode",
+          "input_string": "zabcdefghi",
+          "output_string": "zzzzzzzzzz"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `expected`

File: exercises/simple-cipher/canonical-data.json [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It can handle messages longer than the key",
+          "property": "cipher.key",
+          "cipher_key": "abc",
+          "case": {
+            "property": "encode",
+            "input_string": "iamapandabear",
+            "output_string": "iboaqcnecbfcr"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `expected`

File: exercises/simple-cipher/canonical-data.json [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+          "input_string": "zabcdefghi",
+          "output_string": "zzzzzzzzzz"
+        },
+        {
+          "description": "It can handle messages longer than the key",
+          "property": "cipher.key",
+          "cipher_key": "abc",
+          "case": {
+            "property": "encode",
+            "input_string": "iamapandabear",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It can wrap on decode",
+          "property": "decode",
+          "input_string": "zabcdefghi",
+          "output_string": "zzzzzzzzzz"
+        },
+        {
+          "description": "It can handle messages longer than the key",
+          "property": "cipher.key",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: I would use the `encode` property here and flatten the nested case (with its property removed).

File: exercises/simple-cipher/canonical-data.json [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+          "cipher_key": "abc",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Maybe just use `key` as the property?

File: exercises/simple-cipher/canonical-data.json [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+        {
+          "description": "It can wrap on encode",
+          "property": "encode",
+          "input_string": "zzzzzzzzzz",
+          "output_string": "zabcdefghi"
+        },
+        {
+          "description": "It can wrap on decode",
+          "property": "decode",
+          "input_string": "zabcdefghi",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Renamed to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+          }
+        },
+        {
+          "description": "It can wrap on encode",
+          "property": "encode",
+          "input_string": "zzzzzzzzzz",
+          "output_string": "zabcdefghi"
+        },
+        {
+          "description": "It can wrap on decode",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Maybe shorten to `Decode wraps input` or something like that.

File: exercises/simple-cipher/canonical-data.json [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+          "case": {
+            "property": "encode",
+            "input_string": "iamapandabear",
+            "output_string": "qayaeaagacia"
+          }
+        },
+        {
+          "description": "It can wrap on encode",
+          "property": "encode",
+          "input_string": "zzzzzzzzzz",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+          "input_encode": "abcdefghij",
+          "output_decode": "abcdefghij"
+        },
+        {
+          "description": "It can double shift encode",
+          "property": "key",
+          "cipher_key": "iamapandabear",
+          "case": {
+            "property": "encode",
+            "input_string": "iamapandabear",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+        {
+          "description": "It can decode",
+          "property": "decode",
+          "input_string": "abcdefghij",
+          "output_string": "aaaaaaaaaa"
+        },
+        {
+          "description": "It is reversible",
+          "property": "encode -> decode",
+          "input_encode": "abcdefghij",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+        {
+          "description": "It can encode",
+          "property": "encode",
+          "input_string": "aaaaaaaaaa",
+          "output_key": "abcdefghij"
+        },
+        {
+          "description": "It can decode",
+          "property": "decode",
+          "input_string": "abcdefghij",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+        {
+          "description": "It keeps submitted key",
+          "property": "key",
+          "input_key": "cipher.key",
+          "output_key": "input_key"
+        },
+        {
+          "description": "It can encode",
+          "property": "encode",
+          "input_string": "aaaaaaaaaa",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+          "output_key": "abcdefghij"
+        },
+        {
+          "description": "It can decode",
+          "property": "decode",
+          "input_string": "abcdefghij",
+          "output_string": "aaaaaaaaaa"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `expected`

File: exercises/simple-cipher/canonical-data.json [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It is reversible",
+          "property": "encode -> decode",
+          "input_encode": "abcdefghij",
+          "output_decode": "abcdefghij"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `expected`

File: exercises/simple-cipher/canonical-data.json [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        {
+          "description": "It can double shift encode",
+          "property": "key",
+          "cipher_key": "iamapandabear",
+          "case": {
+            "property": "encode",
+            "input_string": "iamapandabear",
+            "output_string": "qayaeaagacia"
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `expected`

File: exercises/simple-cipher/canonical-data.json [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+        }
+      ]
+    },
+    {
+      "description": "Substitution cipher",
+      "property": "substitution",
+      "cipher_key_string": "abcdefgihj",
+      "cases": [
+        {
+          "description": "It keeps submitted key",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: What does this test do?

File: exercises/simple-cipher/canonical-data.json [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+          "property": "Cipher Key",
+          "input_empty_string": "''",
+          "output_error": "Bad key"
+        }
+      ]
+    },
+    {
+      "description": "Substitution cipher",
+      "property": "substitution",
+      "cipher_key_string": "abcdefgihj",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Is this the key that is used in the nested cases? If so, I would suggest removing it here and copying it to each single test case. I would also rename it to `key`.

File: exercises/simple-cipher/canonical-data.json (hunk 3 above)
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Should this be `encode`?

File: exercises/simple-cipher/canonical-data.json [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+        {
+          "description": "It throws an error with an all caps key",
+          "property": "Cipher Key",
+          "input_all_caps_string": "ABCDEF",
+          "output_error": "Bad key"
+        },
+        {
+          "description": "It throws an error with a numeric key",
+          "property": "Cipher Key",
+          "input_all_number_string": "'12345'",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Rename to `input`

File: exercises/simple-cipher/canonical-data.json [hunk 34]
```diff
@@ -0,0 +1,123 @@
...
+      "cases": [
+        {
+          "description": "It throws an error with an all caps key",
+          "property": "Cipher Key",
+          "input_all_caps_string": "ABCDEF",
+          "output_error": "Bad key"
+        },
+        {
+          "description": "It throws an error with a numeric key",
+          "property": "Cipher Key",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: I think the property should be `key`

File: exercises/simple-cipher/canonical-data.json [hunk 35]
```diff
@@ -0,0 +1,123 @@
...
+        }
+      ]
+    },
+    {
+      "description": "Incorrect key cipher",
+      "property": "incorrect",
+      "cases": [
+        {
+          "description": "It throws an error with an all caps key",
+          "property": "Cipher Key",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: I think the property should be `key`

File: exercises/simple-cipher/canonical-data.json [hunk 36]
```diff
@@ -0,0 +1,123 @@
...
+          "output_decode": "plaintext string"
+        }
+      ]
+    },
+    {
+      "description": "Incorrect key cipher",
+      "property": "incorrect",
+      "cases": [
+        {
+          "description": "It throws an error with an all caps key",
```
Comment (ErikSchierboom) [inline exercises/simple-cipher/canonical-data.json]: Maybe this test should be about testing a string with a single uppercase character, as that is the minimal counter-example.

File: exercises/simple-cipher/canonical-data.json (hunk 3 above)
Comment (coriolinus) [reply exercises/simple-cipher/canonical-data.json]: Possibly it should be `substitution_encode` or `substitutionEncode`, if there are other encoding methods with different types of inputs and outputs.

**Comment by Insti** [comment]
Are you still working on this @aos ? Do you need any help?

**Comment by aos** [comment]
Hi everyone. My apologies -- I've been very busy at work and have not had the time to work on this. If someone else would like to takeover, I'm more than happy to pass it on. Otherwise, I'll try to make progress on it as I collect some free time.

**Comment by ErikSchierboom** [comment]
@aos Just a small bump. Are you still interested in working on this? I totally understand if work has prevented you to work on this, but this is just a small, friendly reminder :)

**Comment by ErikSchierboom** [comment]
What should we do with this PR? It has been 29 days since my reminder, and unfortunately there has been no reply. Is someone else perhaps willing to continue working on this PR?

**Comment by rpottsoh** [comment]
Would this PR simply be closed and a new one opened?  Would it matter if one decided to start over from scratch?  I am not putting my hat in the ring at this point, but I have spent about 90 minutes playing around with the JSON and looking at the recommended changes and finding it difficult to make sense of this JSON and have pass linting after making the prescribed changes.

**Comment by ErikSchierboom** [comment]
That may be the best option indeed.

**Comment by ErikSchierboom** [comment]
I'm gonna go ahead and close this PR. If someone wants to continue working on it, that would be great! But please open another PR.
